src/_pytest/python.py

```diff
@@ -911,6 +912,165 @@ def hasnew(obj: object) -> bool:
+
+        # All IDs must be unique!
+        unique_ids = set(resolved_ids)
+        if len(unique_ids) != len(resolved_ids):
+
+            # Record the number of occurrences of each test ID.
+            test_id_counts = Counter(resolved_ids)
+
+            # Map the test ID to its next suffix.
+            test_id_suffixes: Dict[str, int] = defaultdict(int)
+
+            # Suffix non-unique IDs to make them unique.
+            for index, test_id in enumerate(resolved_ids):
+                if test_id_counts[test_id] > 1:
+                    resolved_ids[index] = f"{test_id}{test_id_suffixes[test_id]}"
+                    test_id_suffixes[test_id] += 1
+
+        return resolved_ids
+
+    def make_parameter_keys(self) -> Iterable[Dict[str, Hashable]]:
+        """Make hashable keys for parameters in a parametrized test.
+
+        This key will be considered to determine if parameters
+        are the same in the sense of reorder_items and the FixtureDef cache.
+        The key is guaranteed to be hashable and comparable.
+        It's not intended for printing and therefore not ASCII escaped.
+        """
+        for valindex, parameterset in enumerate(self.parametersets):
+            if parameterset.id is not None:
+                yield {argname: parameterset.id for argname in self.argnames}
+            else:
+                id = self._id_at_index(valindex)
+                if id is None:
+                    yield self._param_keys_from_parameterset(parameterset, valindex)
+                else:
+                    yield {argname: id for argname in self.argnames}
+
+    def _make_resolved_ids(self) -> Iterable[str]:
+        """Combine multiple parameter IDs to a single string for each parameter set."""
+        for valindex, parameterset in enumerate(self.parametersets):
+            if parameterset.id is not None:
+                yield parameterset.id
+            else:
+                id = self._id_at_index(valindex)
+                if id is None:
+                    yield self._test_id_from_parameterset(parameterset, valindex)
+                else:
+                    yield _ascii_escaped_by_config(id, self.config)
+
+    def _id_at_index(self, valindex: int) -> Optional[str]:
+        if self.ids and valindex < len(self.ids):
+            return self.ids[valindex]
+        return None
+
+    def _test_id_from_parameterset(self, parameterset, valindex) -> str:
+        this_id = [
+            self._idval(val, argname, valindex)
+            for val, argname in zip(parameterset.values, self.argnames)
+        ]
+        return "-".join(this_id)
+
+    def _param_keys_from_parameterset(
+        self, parameterset, valindex
+    ) -> Dict[str, Hashable]:
+        this_id: Dict[str, Hashable] = {}
+        for val, argname in zip(parameterset.values, self.argnames):
+            evaluated_id = self._evaluate_id_function(val, argname, valindex)
+            if isinstance(evaluated_id, str):
```

Comment:
TODO: Support cases where a user provided `id=` function doesn't directly return `str`, but a type that can reasonably converted to `str` (like `Enum`, see `_autoconvert_idval`).